warp10/src/main/java/io/warp10/script/functions/MOTIONSPLIT.java

```diff
@@ -0,0 +1,222 @@
+   * @param proximityZoneSpeed Minimum speed in the proxmityZone to prevent a split, in m/s
+   * 
+   * @return
+   * @throws WarpScriptException
+   */
+  private static List<GeoTimeSerie> motionSplit(GeoTimeSerie gts, String label, long timeThreshold, double distanceThreshold, double proximityZoneRadius, double proximityZoneSpeed, long proximityZoneTime) throws WarpScriptException {
+    //
+    // Sort GTS according to timestamps
+    //
+    
+    GTSHelper.sort(gts, false);
+
+    GeoTimeSerie split = null;
+    
+    List<GeoTimeSerie> splits = new ArrayList<GeoTimeSerie>();
+    
+    int splitStart = 0;
+    int idx = 0;
+    
+    int n = GTSHelper.nvalues(gts);
+    
+    boolean mustSplit = true;
+
+    long refLocation = GeoTimeSerie.NO_LOCATION;
+    long refTimestamp = Long.MIN_VALUE;
+    
+    long previousTimestamp = Long.MIN_VALUE;
+    long previousLocation = GeoTimeSerie.NO_LOCATION;
+    long proximityZoneTraveledDistance = 0;
+    
+    while (idx < n) {
+      long timestamp = GTSHelper.tickAtIndex(gts, idx);
+      long location = GTSHelper.locationAtIndex(gts, idx);
+      long elevation = GTSHelper.elevationAtIndex(gts, idx);
+      Object value = GTSHelper.valueAtIndex(gts, idx);
+    
+      //
+      // If the previous tick was more than 'timeThreashold' ago, split now
+      //
+      
+      if (timestamp - previousTimestamp > timeThreshold) {
+        mustSplit = true;
+      }
+      
+      //
+      // If the distance to the previous location is above distanceThreshold, split
+      //
+      
+      if (GeoTimeSerie.NO_LOCATION != previousLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(location, previousLocation) > distanceThreshold) {
+        mustSplit = true;
+      }
+      
+      //
+      // If the current point is farther away from the reference point than 'proximityDistance'
+      // change the reference point and reset the traveled distance.
+      // If the distance traveled in the proximity zone is less than the proximityTrip and the time spent in the zone is above maxProximityTime force a split
+      //
+      
+      if (GeoTimeSerie.NO_LOCATION != refLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(refLocation, location) > proximityZoneRadius) {
+        if (timestamp - refTimestamp > proximityZoneTime && proximityZoneTraveledDistance / ((timestamp - refTimestamp) / Constants.TIME_UNITS_PER_S) < proximityZoneSpeed) {
+          mustSplit = true;
+        }
+        refLocation = location;
+        refTimestamp = timestamp;
+        proximityZoneTraveledDistance = 0;
+      }
+      
+      if (GeoTimeSerie.NO_LOCATION != refLocation && GeoTimeSerie.NO_LOCATION != location && GeoXPLib.orthodromicDistance(refLocation, location) <= proximityZoneRadius) {
```

Comment:
Should be a `else` because when `orthodromicDistance(refLocation, location) > proximityZoneRadius` is `true` then `refLocation = location` and then `orthodromicDistance(refLocation, location) <= proximityZoneRadius` (well, highly likely, because it is 0) so `orthodromicDistance(previousLocation, location)` is added to `proximityZoneTraveledDistance`.

This means, the next iteration, in the case `orthodromicDistance(refLocation, location) <= proximityZoneRadius`, `proximityZoneTraveledDistance` is also updated but it already contains the distance from the previous iteration although this distance was not traveled inside the proximity zone.